exasol_data_science_utils_python/model_utils/reservoir_shuffle.py

```diff
@@ -0,0 +1,109 @@
+            batch_rows_to_move_to_buffer = batch_df.values[0:elements_to_fill_count]
+            self.buffer_df.values[elements_to_fill_in_buffer] = batch_rows_to_move_to_buffer
+            self.free_bitset[elements_to_fill_in_buffer] = 0
+            self.fill_level += elements_to_fill_count
+            remaining_df = batch_df.iloc[elements_to_fill_count:]
+            return remaining_df
+        else:
+            return batch_df
+
+    def generate_batch(self, remaining_df: pd.DataFrame):
+        result_count = max(1,
+                           len(remaining_df) // self.batch_size + 0 if len(remaining_df) % self.batch_size == 0 else 1)
+        result_size = min(result_count * self.batch_size,self.fill_level)
+        choice_range, choice_range_for_remaining_start = self.compute_choice_ranges(remaining_df)
+
+        choice_in_buffer, choice_in_remaining = self.make_random_choices(
+            choice_range, choice_range_for_remaining_start, result_size)
+
+        result_batch_df = self.create_result_batch_from_choices(
+            choice_in_buffer, choice_in_remaining, remaining_df)
+
+        elements_to_replace_in_buffer = self.move_remaining_rows_to_buffer(
+            choice_in_buffer, choice_in_remaining, remaining_df)
+
+        self.free_buffer(choice_in_buffer, elements_to_replace_in_buffer)
+
+        return result_batch_df
+
+    def free_buffer(self, choice_in_buffer: np.array, elements_to_replace_in_buffer: np.array):
+        elements_to_free = choice_in_buffer[np.logical_not(np.isin(choice_in_buffer, elements_to_replace_in_buffer))]
+        self.free_bitset[elements_to_free] = 1
+        self.fill_level -= len(elements_to_free)
+
+    def move_remaining_rows_to_buffer(self, choice_in_buffer: np.array, choice_in_remaining: np.array,
+                                      remaining_df: pd.DataFrame):
+        elements_to_replace_in_buffer = choice_in_buffer[:len(remaining_df) - len(choice_in_remaining)]
+        remaining_range = np.arange(len(remaining_df),dtype=int)
+        elements_to_move_to_buffer = remaining_range[np.logical_not(np.isin(remaining_range, choice_in_remaining))]
+        self.buffer_df.values[elements_to_replace_in_buffer] = remaining_df.values[elements_to_move_to_buffer]
+        return elements_to_replace_in_buffer
+
+    def create_result_batch_from_choices(self, choice_in_buffer: np.array, choice_in_remaining: np.array,
+                                         remaining_df: pd.DataFrame):
+        result_in_buffer_df = self.buffer_df.iloc[choice_in_buffer]
+        result_in_remaining_df = remaining_df.iloc[choice_in_remaining]
+        result_df = pd.concat([result_in_buffer_df, result_in_remaining_df])
+        return result_df
+
+    def make_random_choices(self, choice_range: np.array, choice_range_for_remaining_start: int, result_size: int):
+        np.random.shuffle(choice_range)
```

Comment:
If you would use DependencyInjection here, you could even test some internals of the algorithm....
for example the UnitTest could provide a _randomizer_ which returns value sin reverse order or something like that.

Reply:
good idea, but I probably need to inject the np.random.shuffle completely then or maybe the whole function make_random_choices